File: mcdowell/python/ch9-03.py
def magicFast(inputs, start, stop):
    if start > stop or start < 0 or stop > len(inputs) - 1:
        return -1

    mid = (start + stop) // 2
    midValue = inputs[mid]

    if midValue == mid:
        return mid

    left = magicFast(inputs, start, min([mid - 1, midValue]))
    if left >= 0:
        return left

    right = magicFast(inputs, max([mid + 1, midValue]), stop)

    return right


inputs = [-10, -10, 0, 3, 3, 3, 3, 10, 11, 12]
print magicFast(inputs, 0, len(inputs) - 1)

File: mcdowell/js/ch9-03.js
function magicFast(inputs, start, stop) {

    if (start > stop) return -1;

    var mid = parseInt((start + stop) / 2);
    var midValue = inputs[mid];

    if (mid == midValue) return mid;

    var left = magicFast(inputs, start, Math.min(midValue, mid - 1));
    if (left != -1) return left;

    var right = magicFast(inputs, Math.max(midValue, mid + 1), stop);

    return right;


}




var inputs = [-10, -10, 0, 3, 3, 3, 3, 10, 11, 12];
console.log( magicFast(inputs, 0, inputs.length - 1)  );